txn.go

```diff
@@ -621,3 +646,46 @@ func (txn *txn) WriteJSON(w io.Writer, tables ...string) error {
 	buf.WriteString("\n}\n")
 	return buf.Flush()
 }
+
+func (txn *txn) gcDeleted(table *tableEntry) {
+	// Look for observed ex-objects and delete them. We look until an object
+	// has higher revision than the highest observed/existing deleted object.
+	endRevision := min(table.deletedRange.end, table.deletedLowWatermark())
+
+	if table.deletedRange.begin >= endRevision {
+		// No deleted objects to check.
+		return
+	}
+
+	primIndexer := table.meta.primary()
+	revTree := txn.mustIndexWriteTxn(table.meta, RevisionIndexPos)
+	primTree := txn.mustIndexWriteTxn(table.meta, PrimaryIndexPos)
+
+	var revKey [8]byte // to avoid heap allocation
+	binary.BigEndian.PutUint64(revKey[:], table.deletedRange.begin+1)
+
+	iter := indexTxn.LowerBound(revTree, revKey[:])
+	for _, obj, ok := iter.Next(); ok; _, obj, ok = iter.Next() {
+		if obj.revision > endRevision {
+			break
+		}
+		if obj.deleted {
+			binary.BigEndian.PutUint64(revKey[:], obj.revision)
+			_, hadOld := revTree.Delete(revKey[:])
+			if !hadOld {
+				panic("BUG: Object to be deleted not found from revision index")
+			}
+			idKey := primIndexer.fromObject(object{data: obj.data}).First()
+			_, hadOld = primTree.Delete(idKey)
+			if !hadOld {
+				panic("BUG: Object to be deleted not found from primary index")
+			}
+			table.numDeletedObjects--
```

Comment:
I guess you could call it `numObjsPendingDeletion` instead, which would look more intuitive here, but it's fine either way